kernel/src/hil/block_storage.rs

```diff
@@ -0,0 +1,247 @@
+/// `D`: The size of a discard block in bytes.
+pub trait WriteableStorage<const W: usize, const D: usize> {
+    /// Write data from a buffer to storage.
+    ///
+    /// This function writes the contents of `buf` to memory,
+    /// at the chosen `region`.
+    ///
+    /// `ErrorCode::INVAL` will be returned if
+    /// - `region` exceeds the end of the device, or
+    /// - `buf` is shorter than `region`.
+    ///
+    /// This function SHALL NOT discard the block first.
+    /// The user of this function MUST ensure that the relevant block
+    /// has been successfully discarded first (see `discard`).
+    ///
+    /// Once a byte has been written as part of a write block,
+    /// it MUST NOT be written again until it's discarded
+    /// as part of a discard block.
+    /// Multiple consecutive writes to the same block
+    /// are forbidden by this trait, but the restriction is not enforced.
+    ///
+    /// **Note** about raw flash devices: writes can turn bits from `1` to `0`.
+    /// To change a bit from `0` to `1`, a region must be erased (discarded).
+    ///
+    /// Returns `ErrorCode::BUSY` when another operation is in progress.
+    ///
+    /// On success, triggers `WriteableClient::write_complete` once.
+    fn write(
+        &self,
+        region: &BlockIndex<W>,
+        buf: &'static mut [u8],
+    ) -> Result<(), (ErrorCode, &'static mut [u8])>;
+
+    /// Makes a region ready for writing.
+    ///
+    /// This corresponds roughly to the erase operation on raw flash devices.
+    ///
+    /// A successful `discard` leaves bytes in the selected `region` undefined.
+    /// The user of this API must not assume any property
+    /// of the discarded bytes.
+    ///
+    /// If `region` exceeds the size of the device, returns `ErrorCode::INVAL`.
+    ///
+    /// Returns `ErrorCode::BUSY` when another operation is in progress.
+    ///
+    /// On success, triggers `WriteableClient::discard_complete` once.
+    fn discard(&self, region: &BlockIndex<D>) -> Result<(), ErrorCode>;
+}
+
+pub trait Storage<const W: usize, const D: usize>:
+    ReadableStorage<W> + WriteableStorage<W, D>
+{
+}
+
+impl<const W: usize, const D: usize, T: ReadableStorage<W> + WriteableStorage<W, D>> Storage<W, D>
+    for T
+{
+}
+
+/// Specifies a storage area with byte granularity.
+pub struct AddressRange {
+    /// Offset from the beginning of the storage device.
+    pub start_address: u64,
+    /// Length of the range.
+    pub length_bytes: u32,
+}
+
+impl AddressRange {
+    pub fn get_end_address(&self) -> u64 {
+        self.start_address + self.length_bytes as u64
+    }
+}
+
+impl<const C: usize> From<BlockIndex<C>> for AddressRange {
+    fn from(region: BlockIndex<C>) -> Self {
+        AddressRange {
+            start_address: region.0 as u64 * C as u64,
+            length_bytes: C as u32,
+        }
+    }
+}
+
+/// Devices which can read arbitrary byte-indexed ranges.
+pub trait ReadRange {
```

Comment:
This trait is for a device?

Also, it seems out of place to have byte level reads in a block storage driver.

Reply:
I am not particularly attached to it, but I've noticed that flash devices include byte addressing mode for reads. Should I get rid of it?

Reply:
What about having `ReadableStorageBlocks` and `ReadableStorageBytes`. Then drivers can implement whatever they support.